vm coredump support

up2wing opened the pull request:
It's useful to dump the guest so that crash tool can be used to analysize it when guest hung up.

The crash tool can analysize the vmcore as elf file. To coredump the guest for crash tool, we should save the cpu registers on elf header and the guest memory as PT_LOAD.

For now crash tool parse Linux version and so on according to the note segment which named 'QEMU', so the easies way to make crash tool work is adding the 'QEMU' note segment.

rbradford commented on vmm/src/vm.rs:
```diff
@@ -2842,6 +2843,31 @@ impl Migratable for Vm {
 }
 
+impl Coredumpable for Vm {
+    fn coredump(&mut self, _destination_url: &str) -> std::result::Result<(), MigratableError> {
```
Could this possibly be part of the `Debuggable` trait?

rbradford commented on vm-migration/src/lib.rs:
```diff
@@ -313,3 +319,10 @@ pub trait Migratable: Send + Pausable + Snapshottable + Transportable {
 }
+
+/// Trait to be implemented by componet that can be coredumped.
+pub trait Coredumpable: Pausable {
```
This doesn't look like the correct crate for this.

rbradford commented on vm-migration/src/lib.rs:
```diff
@@ -313,3 +319,10 @@ pub trait Migratable: Send + Pausable + Snapshottable + Transportable {
     }
 }
+
+/// Trait to be implemented by componet that can be coredumped.
```
nit: typo "component"

up2wing replied:
thanks.

rbradford commented on vm-migration/Cargo.toml:
```diff
@@ -13,3 +13,4 @@ serde_json = "1.0.79"
 versionize = "0.1.6"
 versionize_derive = "0.1.4"
 vm-memory = { version = "0.7.0", features = ["backend-mmap", "backend-atomic"] }
+object = "0.28.3"
```
I don't really want to add a new dependency if we can help it. There all ELF constants available in linux-loader which is something we already depend upon: https://github.com/rust-vmm/linux-loader/blob/main/src/loader_gen/x86_64/elf.rs

rbradford replied:
The `U64` and `U32` types are for endianess. In Cloud Hypervisor we only support LE and will only support LE so I don't think you should worry about that.

up2wing replied:
Well, actually I met a packed problem when saving elf note header and U32 and U64 solved that. Next I will use elf type in linux-loader to try.

up2wing replied:
Emmm, module `loader_gen` is private, we can't use the elf definition in it.
Any suggestions? @rbradford

rbradford replied:
It's `use pub` exposed as `linux_loader::elf` so e.g. `use linux_loader::elf::Elf64_Phdr;`

up2wing replied:
That works, thx.

rbradford commented on vmm/src/api/http.rs:
```diff
@@ -161,6 +161,8 @@ lazy_static! {
         r.routes.insert(endpoint!("/vm.send-migration"), Box::new(VmActionHandler::new(VmAction::SendMigration(Arc::default()))));
         r.routes.insert(endpoint!("/vm.shutdown"), Box::new(VmActionHandler::new(VmAction::Shutdown)));
         r.routes.insert(endpoint!("/vm.snapshot"), Box::new(VmActionHandler::new(VmAction::Snapshot(Arc::default()))));
+        #[cfg(feature = "gdb")]
```
Probably a good idea to rename this as `guest_debug`. @retrage Do you see any problem with that?

retrage replied:
It's reasonable to rename the feature flag as the coredump feature is not a part of GDB support, but it shares debugger-independent code.

up2wing replied:
Do you mean move coredump_xxx from Debuggable trait into a new trait which may be GuestDebuggable? If so, I will move all the new code in gdb.rs into a new separate module coredump.

rbradford replied:
@up2wing There are two things:

1. Using a different name for the feature flag e.g. `#[cfg(feature = "gdb")]`-> `#[cfg(feature = "guest_debug")]` and 
2. Putting your new code in a new module rather than putting it in the gdb module.

liuw replied:
Or use a dedicated `coredump` feature to gate so that users can decide to include one or both.

rbradford commented on vmm/src/gdb.rs:
```diff
@@ -76,6 +128,303 @@ pub trait Debuggable: vm_migration::Pausable {
+}
+
+/// transfer anything to &[u8]
+pub unsafe fn any_as_u8_slice<T: Sized>(p: &T) -> &[u8] {
```
I think we already have `ByteValued` that already does this?

up2wing replied:
ByteValued works find. Thanks.

rbradford commented on vmm/src/gdb.rs:
```diff
@@ -76,6 +128,303 @@ pub trait Debuggable: vm_migration::Pausable {
+}
+
+#[repr(C)]
+pub struct CldhCpuSegment {
```
You don't need to prefix with `Cldh`

up2wing replied:
ok.

rbradford commented on vmm/src/gdb.rs:
```diff
@@ -76,6 +128,303 @@ pub trait Debuggable: vm_migration::Pausable {
+}
+
+#[repr(C)]
+pub struct X86_64UserRegs {
```
Can you put the new code in a separate module e.g. `coredump` ?

rbradford reviewed (commented):
Firstly thank you for contribution. There is a lot of extra code here and I really want to understand if it worth it over e.g. our existing gdb support or the ability to do a snapshot and use the memory from that (a separate tool good be included in the tree to merge those into the ELF file if necessary.)

@up2wing Do you have a specific use case for this approach?

@liuw You opened the original issue and so I need your feedback here.

up2wing commented:
> Firstly thank you for contribution. There is a lot of extra code here and I really want to understand if it worth it over e.g. our existing gdb support or the ability to do a snapshot and use the memory from that (a separate tool good be included in the tree to merge those into the ELF file if necessary.)

Thx for your quick response :-)

Of course using existing code is the best, however, the most of this patch set is saving info about elf file e.g. program header, note header which are not existing code.

As to memory part in snapshot, I did use that in the patch :-)

> 
> @up2wing Do you have a specific use case for this approach?

My fault. I meant to write the doc after the code have been done.

We can use the coredump function like this:

ch-remote --api-socket=/tmp/ch-sock pause
ch-remote --api-socket=/tmp/ch-sock coredump file:///home/coredump/ch.dmp
crash vmlinux ch.dmp

up2wing commented on vm-migration/Cargo.toml:
```diff
@@ -13,3 +13,4 @@ serde_json = "1.0.79"
 versionize = "0.1.6"
 versionize_derive = "0.1.4"
 vm-memory = { version = "0.7.0", features = ["backend-mmap", "backend-atomic"] }
+hypervisor = { path = "../hypervisor" }
```
ok.

up2wing commented:
> @up2wing I think the question about the use case is to understand what `crash` brings to the table in terms of debug. Since Cloud Hypervisor already supports GDB, what is missing that you need and that would explain the reason why you decided to implement the `crash` support?

We can only debug Cloud Hypervisor itself and get registers of guest using GDB,  it's not convenient to debug the real reason when something wrong happens in Gues but Cloud Hypervisor runs normally.

For instance, when deadlock e.g. spinlock, rcu stall happens in guest kernel, we can only watch the vcpu  thread100% in host, but it we want to analysize the reason why deaklock happens, the only way is coredump the guest and using crash tool with vmlinux of guest kernel, 'cause we can't even login the guest now.

up2wing commented:
Hi @rbradford @sboeuf @liuw would you please review the latest patch? :-)

sboeuf commented:
> Hi @rbradford @sboeuf @liuw would you please review the latest patch? :-)

Sure I'll give it a review today.

sboeuf commented on vm-migration/Cargo.toml:
```diff
@@ -6,6 +6,7 @@ edition = "2021"
 
 [dependencies]
 anyhow = "1.0.57"
+hypervisor = { path = "../hypervisor" }
```
I'm concerned about the dependency here. `vm-migration` is a very low level crate and I would expect the `hypervisor` crate to rely on it rather than the other way around.

up2wing replied:
The latest code have already don't need this dependency.

sboeuf commented on src/bin/ch-remote.rs:
```diff
@@ -595,6 +617,15 @@ fn main() {
+                .arg(
+                    Arg::new("coredump_config")
+                        .index(1)
+                        .help("<destination_url>"),
```
I agree, since it's for debug purpose, we'll keep it simple with the coredump generated to a file only.

up2wing replied:
Okay.

sboeuf commented on vmm/src/api/mod.rs:
```diff
@@ -188,6 +191,12 @@ pub struct VmSnapshotConfig {
+#[derive(Clone, Deserialize, Serialize, Default, Debug)]
+pub struct VmCoredumpConfig {
+    /// The coredump destination URL
+    pub destination_url: String,
```
I agree.

sboeuf commented on vmm/src/cpu.rs:
```diff
@@ -44,14 +53,23 @@ use std::os::unix::thread::JoinHandleExt;
     Migratable, MigratableError, Pausable, Snapshot, SnapshotDataSection, Snapshottable,
     Transportable,
 };
+
```
Blank line added, please remove it, as this will prevent `cargo fmt` from organizing the list of imports alphabetically.

up2wing replied:
Okay.

sboeuf commented on vmm/src/lib.rs:
```diff
@@ -561,6 +565,15 @@ impl Vmm {
     }
 
+    #[cfg(feature = "guest_debug")]
+    fn vm_coredump(&mut self, destination_url: &str) -> result::Result<(), VmError> {
```
```suggestion
    fn vm_coredump(&mut self, filepath: &str) -> result::Result<(), VmError> {
```

sboeuf commented on vmm/src/api/http.rs:
```diff
@@ -161,6 +161,8 @@ lazy_static! {
         r.routes.insert(endpoint!("/vm.shutdown"), Box::new(VmActionHandler::new(VmAction::Shutdown)));
         r.routes.insert(endpoint!("/vm.snapshot"), Box::new(VmActionHandler::new(VmAction::Snapshot(Arc::default()))));
+        #[cfg(feature = "guest_debug")]
+        r.routes.insert(endpoint!("/vm.coredump"), Box::new(VmActionHandler::new(VmAction::Coredump(Arc::default()))));
```
Please update the `docs` accordingly so that it reflects the complete API list.

up2wing replied:
Okay.

sboeuf reviewed (changes requested):
I think the current version of your PR is missing the part where you introduce the `coredump` module as part of the `vmm` crate. Could you please add it so that it can be reviewed?
Also, please make sure to rebase your PR on the latest main branch.
And last thing, please reorganize your commits so that you integrate the feedback you received so far directly into each commit (don't add more commits to integrate the feedback).

up2wing commented:
> I think the current version of your PR is missing the part where you introduce the `coredump` module as part of the `vmm` crate. Could you please add it so that it can be reviewed? Also, please make sure to rebase your PR on the latest main branch. And last thing, please reorganize your commits so that you integrate the feedback you received so far directly into each commit (don't add more commits to integrate the feedback).

Thx for the suggestions.

I will rebase on the latest main branch and reorganize all the commits based on the feedback.

up2wing commented on vmm/src/memory_manager.rs:
```diff
@@ -1845,6 +1852,86 @@ impl MemoryManager {
     }
+
+    #[cfg(feature = "guest_debug")]
+    pub fn coredump_get_mapping_num(&self) -> u32 {
```
Okay.

up2wing commented on vmm/src/memory_manager.rs:
```diff
@@ -44,6 +48,9 @@ use vm_migration::{
     Snapshot, SnapshotDataSection, Snapshottable, Transportable, VersionMapped,
 };
+#[cfg(feature = "guest_debug")]
+use crate::coredump::{MemoryManagerCoredumpData, MemCoredumpFileInfo, DumpState};
```
Okay.

up2wing commented:
> > I think the current version of your PR is missing the part where you introduce the `coredump` module as part of the `vmm` crate. Could you please add it so that it can be reviewed? Also, please make sure to rebase your PR on the latest main branch. And last thing, please reorganize your commits so that you integrate the feedback you received so far directly into each commit (don't add more commits to integrate the feedback).
> 
> Thx for the suggestions.
> 
> I will rebase on the latest main branch and reorganize all the commits based on the feedback.

@sboeuf 
Since the code have been modified a few versions, would I open a new PR? If no, how should I do? Please forgive me not familiar on github -_-

rbradford commented:
> > > I think the current version of your PR is missing the part where you introduce the `coredump` module as part of the `vmm` crate. Could you please add it so that it can be reviewed? Also, please make sure to rebase your PR on the latest main branch. And last thing, please reorganize your commits so that you integrate the feedback you received so far directly into each commit (don't add more commits to integrate the feedback).
> > 
> > 
> > Thx for the suggestions.
> > I will rebase on the latest main branch and reorganize all the commits based on the feedback.
> 
> @sboeuf Since the code have been modified a few versions, would I open a new PR? If no, how should I do? Please forgive me not familiar on github -_-

The best solution is to use `git rebase -i` to rearrange your work and then force push to your branch which will update the PR.

up2wing commented:
> > > > I think the current version of your PR is missing the part where you introduce the `coredump` module as part of the `vmm` crate. Could you please add it so that it can be reviewed? Also, please make sure to rebase your PR on the latest main branch. And last thing, please reorganize your commits so that you integrate the feedback you received so far directly into each commit (don't add more commits to integrate the feedback).
> > > 
> > > 
> > > Thx for the suggestions.
> > > I will rebase on the latest main branch and reorganize all the commits based on the feedback.
> > 
> > 
> > @sboeuf Since the code have been modified a few versions, would I open a new PR? If no, how should I do? Please forgive me not familiar on github -_-
> 
> The best solution is to use `git rebase -i` to rearrange your work and then force push to your branch which will update the PR.

Thx for your quick response. I will try.

liuw commented:
@up2wing you should use `git rebase -i` to rebase your patches on top of the main branch. You did a merge instead. That's not the correct workflow.

up2wing commented:
Hi all,

I used `rebase -i` to reorganize the work, and then fetch and merge upstream/main, but the PR seems a terrible mess now, on this I feel so sorry. Can you see the changes I made? How should I do now?
@sboeuf @rbradford @liuw

liuw commented:
> Hi all,
> 
> I used `rebase -i` to reorganize the work, and then fetch and merge upstream/main, but the PR seems a terrible mess now, on this I feel so sorry. Can you see the changes I made? How should I do now? @sboeuf @rbradford @liuw 

The correct sequence should be

* Fetch from upstream main
* On your topic branch, run `git rebase -i upstream/main` (assuming your use upstream as the remote name)
  * Make sure the list of commits are the ones you want to upstream.
 
There is no need to merge at all.

And don't panic. It can be fixed. :-)

up2wing commented:
> > Hi all,
> > I used `rebase -i` to reorganize the work, and then fetch and merge upstream/main, but the PR seems a terrible mess now, on this I feel so sorry. Can you see the changes I made? How should I do now? @sboeuf @rbradford @liuw
> 
> The correct sequence should be
> 
> * Fetch from upstream main
> * On your topic branch, run `git rebase -i upstream/main` (assuming your use upstream as the remote name)
>   
>   * Make sure the list of commits are the ones you want to upstream.

Many thx, I will follow this step in the future. 

> There is no need to merge at all.
> 
> And don't panic. It can be fixed. :-)

As to this PR, what can I do now?

liuw commented:
> > There is no need to merge at all.
> > And don't panic. It can be fixed. :-)
> 
> As to this PR, what can I do now?

If you follow my steps above, the history of this `coredump` branch will be clean and linear. You can then do a force push. GitHub should do the right thing all by itself.

up2wing commented:
> > > There is no need to merge at all.
> > > And don't panic. It can be fixed. :-)
> > 
> > 
> > As to this PR, what can I do now?
> 
> If you follow my steps above, the history of this `coredump` branch will be clean and linear. You can then do a force push. GitHub should do the right thing all by itself.

I followed steps you said, there are some conflicts during `git rebase -i`, but those are not my code, so I `git rebase --skip`.

I am not sure what I did is totally right, please review again. Many thx.

liuw commented:
> 



> > > > There is no need to merge at all.
> > > > And don't panic. It can be fixed. :-)
> > > 
> > > 
> > > As to this PR, what can I do now?
> > 
> > 
> > If you follow my steps above, the history of this `coredump` branch will be clean and linear. You can then do a force push. GitHub should do the right thing all by itself.
> 
> I followed steps you said, there are some conflicts during `git rebase -i`, but those are not my code, so I `git rebase --skip`.
> 
> I am not sure what I did is totally right, please review again. Many thx.

No, don't skip. You should fix the conflicts. Git stops at the (your) commit that introduces a conflict.

up2wing commented:
> > 
> 
> > > > > There is no need to merge at all.
> > > > > And don't panic. It can be fixed. :-)
> > > > 
> > > > 
> > > > As to this PR, what can I do now?
> > > 
> > > 
> > > If you follow my steps above, the history of this `coredump` branch will be clean and linear. You can then do a force push. GitHub should do the right thing all by itself.
> > 
> > 
> > I followed steps you said, there are some conflicts during `git rebase -i`, but those are not my code, so I `git rebase --skip`.
> > I am not sure what I did is totally right, please review again. Many thx.
> 
> No, don't skip. You should fix the conflicts. Git stops at the (your) commit that introduces a conflict.

Those conflicts are not mine, should I fix them when `git rebase -i upstream/main`?
Now I have pushed the new commits to my branch, what can I do now? :sweat_smile:

liuw commented:
> > > 
> > 
> > 
> > > > > > There is no need to merge at all.
> > > > > > And don't panic. It can be fixed. :-)
> > > > > 
> > > > > 
> > > > > As to this PR, what can I do now?
> > > > 
> > > > 
> > > > If you follow my steps above, the history of this `coredump` branch will be clean and linear. You can then do a force push. GitHub should do the right thing all by itself.
> > > 
> > > 
> > > I followed steps you said, there are some conflicts during `git rebase -i`, but those are not my code, so I `git rebase --skip`.
> > > I am not sure what I did is totally right, please review again. Many thx.
> > 
> > 
> > No, don't skip. You should fix the conflicts. Git stops at the (your) commit that introduces a conflict.
> 
> Those conflicts are not mine, should I fix them when `git rebase -i upstream/main`? Now I have pushed the new commits to my branch, what can I do now? 😅

I now understand what you meant after looking at your branch. It seems that your branch was messed up. The first three patches are already in upstream/main. You can just drop them.

After doing that, your first commit is https://github.com/up2wing/cloud-hypervisor/commit/770e428e70598cca0ce15ede19e2fc4f8d8fd5da. That looks like a squashed commit for multiple commits. I'm not sure if that is your most up-to-date code or not, but that commit should be split into multiple ones.

liuw commented:
And if you've somehow overwrote the history without any backup branch made, you can try using `git reflog` to revive a deleted branch (this depends on Git not having GC'ed unreachable objects, which should rarely happen with a repository like Cloud Hypervisor).

rbradford commented on vmm/src/coredump.rs:
```diff
@@ -0,0 +1,366 @@
+// Copyright 2022 ZTE Corporation.
+// Portions Copyright 2020 The Chromium OS Authors. All rights reserved.
```
Did you import this code from somewhere else? If you wrote it from scratch it should be Apache 2.0 and just the one copyright holder.

up2wing replied:
I wrote this from scratch. I will rewrite this part.

up2wing commented:
> > > > 
> > > 
> > > 
> > > > > > > There is no need to merge at all.
> > > > > > > And don't panic. It can be fixed. :-)
> > > > > > 
> > > > > > 
> > > > > > As to this PR, what can I do now?
> > > > > 
> > > > > 
> > > > > If you follow my steps above, the history of this `coredump` branch will be clean and linear. You can then do a force push. GitHub should do the right thing all by itself.
> > > > 
> > > > 
> > > > I followed steps you said, there are some conflicts during `git rebase -i`, but those are not my code, so I `git rebase --skip`.
> > > > I am not sure what I did is totally right, please review again. Many thx.
> > > 
> > > 
> > > No, don't skip. You should fix the conflicts. Git stops at the (your) commit that introduces a conflict.
> > 
> > 
> > Those conflicts are not mine, should I fix them when `git rebase -i upstream/main`? Now I have pushed the new commits to my branch, what can I do now? sweat_smile
> 
> I now understand what you meant after looking at your branch. It seems that your branch was messed up. The first three patches are already in upstream/main. You can just drop them.
> 

Thanks for your reply. 

> After doing that, your first commit is [up2wing@770e428](https://github.com/up2wing/cloud-hypervisor/commit/770e428e70598cca0ce15ede19e2fc4f8d8fd5da). That looks like a squashed commit for multiple commits. I'm not sure if that is your most up-to-date code or not, but that commit should be split into multiple ones.

I am working on this and patches will come soon.

up2wing commented:
> > > > 
> > > 
> > > 
> > > > > > > There is no need to merge at all.
> > > > > > > And don't panic. It can be fixed. :-)
> > > > > > 
> > > > > > 
> > > > > > As to this PR, what can I do now?
> > > > > 
> > > > > 
> > > > > If you follow my steps above, the history of this `coredump` branch will be clean and linear. You can then do a force push. GitHub should do the right thing all by itself.
> > > > 
> > > > 
> > > > I followed steps you said, there are some conflicts during `git rebase -i`, but those are not my code, so I `git rebase --skip`.
> > > > I am not sure what I did is totally right, please review again. Many thx.
> > > 
> > > 
> > > No, don't skip. You should fix the conflicts. Git stops at the (your) commit that introduces a conflict.
> > 
> > 
> > Those conflicts are not mine, should I fix them when `git rebase -i upstream/main`? Now I have pushed the new commits to my branch, what can I do now? sweat_smile
> 
> I now understand what you meant after looking at your branch. It seems that your branch was messed up. The first three patches are already in upstream/main. You can just drop them.
> 

Thanks for your reply! I `rebase -i` to split the main commit into multiple ones, and
`git fetch  upstream`
`git rebase -i upstream/main`
`git push origin coredump:coredump -f`

but there are still two commit b27da7c, 2766bcd. How to drop them please?

> After doing that, your first commit is [up2wing@770e428](https://github.com/up2wing/cloud-hypervisor/commit/770e428e70598cca0ce15ede19e2fc4f8d8fd5da). That looks like a squashed commit for multiple commits. I'm not sure if that is your most up-to-date code or not, but that commit should be split into multiple ones.

I am working on this and patches will come soon.

sboeuf commented:
> > > > > 
> > > > 
> > > > 
> > > > > > > > There is no need to merge at all.
> > > > > > > > And don't panic. It can be fixed. :-)
> > > > > > > 
> > > > > > > 
> > > > > > > As to this PR, what can I do now?
> > > > > > 
> > > > > > 
> > > > > > If you follow my steps above, the history of this `coredump` branch will be clean and linear. You can then do a force push. GitHub should do the right thing all by itself.
> > > > > 
> > > > > 
> > > > > I followed steps you said, there are some conflicts during `git rebase -i`, but those are not my code, so I `git rebase --skip`.
> > > > > I am not sure what I did is totally right, please review again. Many thx.
> > > > 
> > > > 
> > > > No, don't skip. You should fix the conflicts. Git stops at the (your) commit that introduces a conflict.
> > > 
> > > 
> > > Those conflicts are not mine, should I fix them when `git rebase -i upstream/main`? Now I have pushed the new commits to my branch, what can I do now? sweat_smile
> > 
> > 
> > I now understand what you meant after looking at your branch. It seems that your branch was messed up. The first three patches are already in upstream/main. You can just drop them.
> 
> Thanks for your reply! I `rebase -i` to split the main commit into multiple ones, and `git fetch upstream` `git rebase -i upstream/main` `git push origin coredump:coredump -f`
> 
> but there are still two commit [b27da7c](https://github.com/cloud-hypervisor/cloud-hypervisor/commit/b27da7c71419f167ce2b32c3ea8b67675a0ec44e), [2766bcd](https://github.com/cloud-hypervisor/cloud-hypervisor/commit/2766bcd0efc7d3f8a28f952aa5aebf5f509a1b05). How to drop them please?
> 
> > After doing that, your first commit is [up2wing@770e428](https://github.com/up2wing/cloud-hypervisor/commit/770e428e70598cca0ce15ede19e2fc4f8d8fd5da). That looks like a squashed commit for multiple commits. I'm not sure if that is your most up-to-date code or not, but that commit should be split into multiple ones.
> 
> I am working on this and patches will come soon.

Looking at the patches you've just posted, the only thing you need to do now is:
```bash
git fetch origin
git rebase origin/main
```
And then you push to your branch again.

up2wing commented:
> > > > > > 
> > > > > 
> > > > > 
> > > > > > > > > There is no need to merge at all.
> > > > > > > > > And don't panic. It can be fixed. :-)
> > > > > > > > 
> > > > > > > > 
> > > > > > > > As to this PR, what can I do now?
> > > > > > > 
> > > > > > > 
> > > > > > > If you follow my steps above, the history of this `coredump` branch will be clean and linear. You can then do a force push. GitHub should do the right thing all by itself.
> > > > > > 
> > > > > > 
> > > > > > I followed steps you said, there are some conflicts during `git rebase -i`, but those are not my code, so I `git rebase --skip`.
> > > > > > I am not sure what I did is totally right, please review again. Many thx.
> > > > > 
> > > > > 
> > > > > No, don't skip. You should fix the conflicts. Git stops at the (your) commit that introduces a conflict.
> > > > 
> > > > 
> > > > Those conflicts are not mine, should I fix them when `git rebase -i upstream/main`? Now I have pushed the new commits to my branch, what can I do now? sweat_smile
> > > 
> > > 
> > > I now understand what you meant after looking at your branch. It seems that your branch was messed up. The first three patches are already in upstream/main. You can just drop them.
> > 
> > 
> > Thanks for your reply! I `rebase -i` to split the main commit into multiple ones, and `git fetch upstream` `git rebase -i upstream/main` `git push origin coredump:coredump -f`
> > but there are still two commit [b27da7c](https://github.com/cloud-hypervisor/cloud-hypervisor/commit/b27da7c71419f167ce2b32c3ea8b67675a0ec44e), [2766bcd](https://github.com/cloud-hypervisor/cloud-hypervisor/commit/2766bcd0efc7d3f8a28f952aa5aebf5f509a1b05). How to drop them please?
> > > After doing that, your first commit is [up2wing@770e428](https://github.com/up2wing/cloud-hypervisor/commit/770e428e70598cca0ce15ede19e2fc4f8d8fd5da). That looks like a squashed commit for multiple commits. I'm not sure if that is your most up-to-date code or not, but that commit should be split into multiple ones.
> > 
> > 
> > I am working on this and patches will come soon.
> 
> Looking at the patches you've just posted, the only thing you need to do now is:
> 
> ```shell
> git fetch upstream
> git rebase upstream/main
> git push origin HEAD:coredump --force-with-lease
> ```

I followed your steps but git says everything is up-to-date and there is nothing happened in github :(

```shell
# git rebase -i upstream/main
Successfully rebased and updated refs/heads/coredump.
# git rebase upstream/main 
Current branch coredump is up to date.
# git push origin HEAD:coredump --force-with-lease
Everything up-to-date
```

sboeuf commented:
Let me have a quick look.

sboeuf commented:
@up2wing I've fixed it. Now make sure you do:
```
git fetch origin
git reset --hard origin/coredump
```
This way you will rely locally on what I've just updated.

up2wing commented:
> @up2wing I've fixed it. Now make sure you do:
> 
> ```
> git fetch origin
> git reset --hard origin/coredump
> ```
> 
> This way you will rely locally on what I've just updated.

Thanks so much, Sebastien! @sboeuf 

Could you please tell me what's wrong what I did and how did you fix that?

sboeuf commented:
> > @up2wing I've fixed it. Now make sure you do:
> > ```
> > git fetch origin
> > git reset --hard origin/coredump
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This way you will rely locally on what I've just updated.
> 
> Thanks so much, Sebastien! @sboeuf
> 
> Could you please tell me what's wrong what I did and how did you fix that?

Oh yes of course. You had two commits that were introduced earlier in the codebase, but since the second commit was a revert of the first one, Git wasn't complaining about the fact that you were introducing these commits again. So I manually removed them with a `git rebase -i origin/main`.

up2wing commented:
> > > @up2wing I've fixed it. Now make sure you do:
> > > ```
> > > git fetch origin
> > > git reset --hard origin/coredump
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This way you will rely locally on what I've just updated.
> > 
> > 
> > Thanks so much, Sebastien! @sboeuf
> > Could you please tell me what's wrong what I did and how did you fix that?
> 
> Oh yes of course. You had two commits that were introduced earlier in the codebase, but since the second commit was a revert of the first one, Git wasn't complaining about the fact that you were introducing these commits again. So I manually removed them with a `git rebase -i origin/main`.

Got it. Thanks again.

Unfortunately there are some cargo fmt error which I didn't pay attention to and I will fix that.

liuw commented:
```
Digest: sha256:554a795f2438aed04e4b039854bcb24ff5793e851f13a5a4864a833eda84a525
Status: Downloaded newer image for openapitools/openapi-generator-cli:latest
Validating spec (/local/vmm/src/api/openapi/cloud-hypervisor.yaml)
Errors:
	- attribute
	  paths.'/vm.coredump'(put).requestBody.content.'application/json'.schema.#/components/schemas/VmCoredumpData
	  is missing

[error] Spec has 1 errors.
```

up2wing commented:
> ```
> Digest: sha256:554a795f2438aed04e4b039854bcb24ff5793e851f13a5a4864a833eda84a525
> Status: Downloaded newer image for openapitools/openapi-generator-cli:latest
> Validating spec (/local/vmm/src/api/openapi/cloud-hypervisor.yaml)
> Errors:
> 	- attribute
> 	  paths.'/vm.coredump'(put).requestBody.content.'application/json'.schema.#/components/schemas/VmCoredumpData
> 	  is missing
> 
> [error] Spec has 1 errors.
> ```

Thx, I have fixed this.

sboeuf commented:
@up2wing I'll review this tomorrow!

sboeuf commented on vmm/src/memory_manager.rs:
```diff
@@ -1845,6 +1849,38 @@ impl MemoryManager {
+
+    #[cfg(feature = "guest_debug")]
+    pub fn coredump_data(&self, mem_offset: u64) -> MemoryCoredumpData {
+        let mut mem_blocks = BTreeMap::new();
+        let mut offset = mem_offset;
+        for guest_ram_mapping in &self.guest_ram_mappings {
+            let load_info = MemCoredumpFileInfo {
+                offset: 0,
+                filesz: 0,
+                size: guest_ram_mapping.size,
+            };
+            mem_blocks.insert(guest_ram_mapping.gpa, load_info);
+        }
+
+        let mut ram_maps = BTreeMap::new();
+        for (gpa, load) in &mem_blocks {
+            let load_info = MemCoredumpFileInfo {
+                offset,
+                filesz: gpa + load.size,
+                size: load.size,
+            };
+
+            ram_maps.insert(*gpa, load_info);
+            offset += load.size;
+        }
```
Can't you simplify this whole block by going through the list of `guest_ram_mappings` only one time?

sboeuf commented on vmm/src/coredump.rs:
```diff
@@ -0,0 +1,86 @@
+#[derive(Clone)]
+pub struct MemCoredumpFileInfo {
+    pub offset: u64,
+    pub filesz: u64,
```
Could you please rename this field in order to make it more comprehensive?

up2wing replied:
Okay.

sboeuf commented on vmm/src/memory_manager.rs:
```diff
@@ -1845,6 +1849,38 @@ impl MemoryManager {
+    }
+
+    #[cfg(feature = "guest_debug")]
+    pub fn coredump_data(&self, mem_offset: u64) -> MemoryCoredumpData {
```
```suggestion
    pub fn coredump_memory_regions(&self, mem_offset: u64) -> CoredumpMemoryRegions {
```

sboeuf commented on vm-migration/src/lib.rs:
```diff
@@ -59,6 +59,12 @@ pub enum MigratableError {
     #[error("Failed to complete migration for migratable component: {0}")]
     CompleteMigration(#[source] anyhow::Error),
+
+    #[error("Failed to coredump coredumpable component: {0}")]
+    Coredump(#[source] anyhow::Error),
+
+    #[error("File error: {0}")]
+    CoredumpFile(#[source] std::io::Error),
```
I think you can remove these since you added them as part of `GuestDebuggableError`.

up2wing replied:
Right.

sboeuf reviewed (commented):
Nice PR!
- Do you think you could add an integration test to cover this functionality?
- You need to add the `guest_debug` feature to the build matrix (Github Actions) as we want to make sure the build won't get broken.

up2wing commented:
> * build matrix



> Nice PR!
> 
> * Do you think you could add an integration test to cover this functionality?

Okay. I suppose I could have a try.

> * You need to add the `guest_debug` feature to the build matrix (Github Actions) as we want to make sure the build won't get broken.

Okay.

up2wing commented on vmm/src/coredump.rs:
```diff
@@ -0,0 +1,86 @@
+/// Architecture Infomation when coredump.
+#[derive(Default)]
+pub struct DumpState {
+    pub note_size: isize,
+    pub phdr_num: u16,
+    pub sh_info: u32,
```
Okay.

up2wing commented on vmm/src/coredump.rs:
```diff
@@ -0,0 +1,86 @@
+    pub file: Option<File>,
+}
+
+impl DumpState {
+    /// create an empty DumpState.
+    pub fn new(
+        note_size: isize,
+        phdr_num: u16,
+        sh_info: u32,
+        mem_offset: u64,
+        mem_info: Option<MemoryCoredumpData>,
+        file: Option<File>,
+    ) -> Self {
+        DumpState {
+            note_size,
+            phdr_num,
+            sh_info,
+            mem_offset,
+            mem_info,
+            file,
+        }
+    }
+}
```
Right.

up2wing commented on vmm/src/coredump.rs:
```diff
@@ -0,0 +1,86 @@
+}
+
+pub trait GuestDebuggable: vm_migration::Pausable {
+    fn coredump(
+        &mut self,
+        _destination_url: &str,
+    ) -> std::result::Result<(), GuestDebuggableError> {
+        Ok(())
+    }
+    fn coredump_init(
+        &mut self,
+        _destination_url: &str,
+    ) -> std::result::Result<DumpState, GuestDebuggableError> {
+        Ok(DumpState::new(0, 0, 0, 0, None, None))
+    }
+    fn coredump_begin(
+        &mut self,
+        _dump_state: &DumpState,
+    ) -> std::result::Result<(), GuestDebuggableError> {
+        Ok(())
+    }
+    fn coredump_iterate(
+        &mut self,
+        _dump_state: &DumpState,
+    ) -> std::result::Result<(), GuestDebuggableError> {
+        Ok(())
+    }
+    fn get_mem_offset(&self, _phdr_num: u16, _note_size: isize) -> u64 {
+        0
+    }
```
It looks clean to simplify this trait only one coredump() method, but if we do that, there may be more internal functions in Vm/MemoryManager, which seem like:
impl Vm/MemoryManager {
    fn coredump_init() {}
    fn coredump_begin() {}
}

yet they belong to GuestDebuggable trait.

sboeuf replied:
Yes and that's fine to introduce more internal functions. A trait is designed from a consumer perspective, and in that case we want to generate a coredump. The internal complexity should be left to each implementation.

up2wing replied:
Hi Sebastien,
Sorry to update this PR so late, 'cause I have some other work to do so it's late here.
I looked through other traits, they generally do the similar functions in one trait and implement them in places where needs to be done:
start/stop_dirty_log() in Migratable trait, debug_pause/resume() in Debuggable trait and so on.
So I prefer the current code in GuestDebuggable trait, what's your opinion? :-)

sboeuf replied:
The question is not about how many functions we have per trait on average, but if they clearly define different operations.
The trait you've defined here is a bit confusing. Just looking at the trait functions, I have no clue what are the differences between them. From my understanding, we want each component implementing the trait to provide a way to generate a coredump, and I thought that having a single method called `coredump()` would be much clearer.
The internals about how the coredump is performed is specific to each implementation, and that's why I don't see any added value in having a trait with so many methods.

up2wing replied:
Thanks for the suggestion. I will look through the code and make some refactors.

up2wing replied:
The code have been refactored, please review again. @sboeuf

up2wing commented on vmm/src/memory_manager.rs:
```diff
@@ -1883,6 +1885,58 @@ impl MemoryManager {
+        &mut self,
+        _dump_state: &DumpState,
+    ) -> std::result::Result<(), GuestDebuggableError> {
+        let memory_ranges = self
+            .memory_range_table(true)
+            .map_err(|e| GuestDebuggableError::Coredump(e.into()))?;
+
+        self.snapshot_memory_ranges = memory_ranges;
```
Of cource we can do this in coredump_iterate(). 
However, if I understand correctly, in general when we design a work flow of feature, we do some initial work in xxx_begin() and do the real job in xxx_start() and so on. So I suppose this also can make sense :-)

up2wing commented on vmm/src/coredump.rs:
```diff
@@ -266,11 +361,31 @@ pub trait Elf64Writable {
     fn get_note_size(&self, desc_type: NoteDescType, nr_cpus: u32) -> u32 {
         let note_head_size = std::mem::size_of::<elf::Elf64_Nhdr>() as u32;
         let elf_desc_size = std::mem::size_of::<X86_64ElfPrStatus>() as u32;
+        let cldh_desc_size = std::mem::size_of::<CldhCpuState>() as u32;
```
Okay~

up2wing commented on vmm/src/cpu.rs:
```diff
@@ -238,6 +255,13 @@ struct InterruptSourceOverride {
 }
 
+#[cfg(feature = "guest_debug")]
+macro_rules! round_up {
```
Indeed there is only one `round_up` macro, there is another  similar macro which names `div_round_up`.

sboeuf replied:
My bad, I didn't realize these were two different macros.

sboeuf commented on vmm/src/memory_manager.rs:
```diff
@@ -1846,6 +1852,88 @@ impl MemoryManager {
+    pub fn coredump_memory_regions(&self, mem_offset: u64) -> CoredumpMemoryRegions {
+        let mut mem_blocks = BTreeMap::new();
+        let mut mem_offset_in_elf = mem_offset;
+        for guest_ram_mapping in &self.guest_ram_mappings {
+            let load_info = CoredumpMemoryRegion {
+                mem_offset_in_elf: 0,
+                mem_size: guest_ram_mapping.size,
+            };
+            mem_blocks.insert(guest_ram_mapping.gpa, load_info);
+        }
+
+        let mut ram_maps = BTreeMap::new();
+        for (gpa, load) in &mem_blocks {
+            let load_info = CoredumpMemoryRegion {
+                mem_offset_in_elf,
+                mem_size: load.mem_size,
+            };
+
+            ram_maps.insert(*gpa, load_info);
+            mem_offset_in_elf += load.mem_size;
+        }
```
Ok I understand that you need the list to be sorted by GPA. But that's something that's not obvious, which is why I suggest the following code (naming the variables correctly):
```rust
let mut mappings_sorted_by_gpa = self.guest_ram_mappings.clone();
mappings_sorted_by_gpa.sort_by_key(|m| m.gpa);

let mut mem_offset_in_elf = mem_offset;
let mut ram_maps = BTreeMap::new();
for mapping in mappings_sorted_by_gpa.iter() {
    ram_maps.insert(mapping.gpa, CoredumpMemoryRegion {
        mem_offset_in_elf,
        mem_size: mapping.size,
    });
    mem_offset_in_elf += mapping.size;
}
```

up2wing replied:
Nice implementation! Thanks.

sboeuf commented on vmm/src/coredump.rs:
```diff
@@ -3,6 +3,10 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
+#[cfg(target_arch = "x86_64")]
```
This is a comment about the commit message, you should remove `Co-authored-by: Sebastien Boeuf <sebastien.boeuf@intel.com>` as I don't think I did anything on this commit.

up2wing replied:
Okay.
The reason is maybe there are some code are removed in some rebase work but I kept the commit message.

sboeuf commented on vmm/src/coredump.rs:
```diff
@@ -82,12 +88,101 @@ pub struct X86_64ElfPrStatus {
+    pub base: u64,
+}
+
+pub const DESC_TYPE_SHIFT: u32 = 8;
+pub const DESC_S_SHIFT: u32 = 12;
+pub const DESC_DPL_SHIFT: u32 = 13;
+pub const DESC_P_SHIFT: u32 = 15;
+pub const DESC_P_MASK: u32 = 1 << DESC_P_SHIFT;
+pub const DESC_AVL_SHIFT: u32 = 20;
+pub const DESC_AVL_MASK: u32 = 1 << DESC_AVL_SHIFT;
+pub const DESC_L_SHIFT: u32 = 21;
+pub const DESC_B_SHIFT: u32 = 22;
+pub const DESC_S_MASK: u32 = 1 << DESC_S_SHIFT;
+pub const DESC_G_SHIFT: u32 = 23;
+pub const DESC_G_MASK: u32 = 1 << DESC_G_SHIFT;
```
Same here, do you need all these constants to be made public?

up2wing replied:
Done as the same previous comment.

sboeuf commented on .github/workflows/build.yaml:
```diff
@@ -41,6 +41,9 @@ jobs:
       - name: Build (default features)
         run: cargo rustc --locked --bin cloud-hypervisor -- -D warnings
 
+      - name: Build (default features + guest_debug)
+        run: cargo rustc --locked --bin cloud-hypervisor --features "guest_debug" -- -D warnings
+
```
Looks good, but could you please move this after the `gdb` one?

up2wing replied:
Okay.

sboeuf commented on tests/integration.rs:
```diff
@@ -5664,6 +5669,46 @@ mod parallel {
+            ));
+
+            // Wait to make sure the coredump is completed
+            thread::sleep(std::time::Duration::new(30, 0));
```
Do you really need to sleep here? I'm asking because I would expect the `ch-remote` command to be blocking until the coredump is complete.

up2wing replied:
Oh yes, sleep is not needed here.
Fixed.

sboeuf commented on tests/integration.rs:
```diff
@@ -5664,6 +5669,46 @@ mod parallel {
+
+            assert!(remote_command(&api_socket, "pause", None));
+
+            assert!(remote_command(
+                &api_socket,
+                "coredump",
+                Some(format!("file://{}", vmcore_file).as_str()),
+            ));
```
That's a first step to check if `ch-remote` returns an error, but I also think we could go one step further and maybe check the header of the generated file, along with the file size which we should know.
@rbradford wdyt?

sboeuf reviewed (commented):
Looking good! I still have some comments but the PR is closer from being merged now.

up2wing commented on vmm/src/vm.rs:
```diff
@@ -2955,6 +3017,36 @@ impl Debuggable for Vm {
+
+        let coredump_state = self.get_dump_state(destination_url)?;
+
+        #[cfg(feature = "tdx")]
+        {
+            if self.config.lock().unwrap().tdx.is_some() {
+                return Err(GuestDebuggableError::Coredump(anyhow!(
+                    "Coredump not possible with TDX VM"
+                )));
+            }
+        }
+
+        let current_state = self.get_state().unwrap();
+        if current_state != VmState::Paused {
+            return Err(GuestDebuggableError::Coredump(anyhow!(
+                "Trying to coredump while VM is running"
+            )));
+        }
```
Oh yes, you are right.

up2wing commented on vmm/src/coredump.rs:
```diff
@@ -41,3 +44,190 @@ pub trait GuestDebuggable: vm_migration::Pausable {
+/// Current ELF version.
+pub const EV_CURRENT: u8 = 1;
+/// AMD x86-64 architecture
+pub const EM_X86_64: u16 = 62;
```
Most of the structures are used in cpu.rs and vm.rs, some of the constants are used only in this file, which I'll fix.

up2wing commented on vmm/src/vm.rs:
```diff
@@ -3052,6 +3050,15 @@ impl GuestDebuggable for Vm {
+        self.cpu_manager
+            .lock()
+            .unwrap()
+            .cpu_write_cldh_note(&coredump_state)?;
```
My fault, left some `cldh` code not renamed.

sboeuf reviewed (commented):
Thanks @up2wing the PR looks good to me. Last thing is about the integration test that could be slightly extended to test a bit more than just if `ch-remote` returns an error or not.

@rbradford PTAL

up2wing commented:
> Thanks @up2wing the PR looks good to me. Last thing is about the integration test that could be slightly extended to test a bit more than just if `ch-remote` returns an error or not.
> 
> @rbradford PTAL

Okay, I will add some more checks using readelf.

up2wing commented:
> Thanks @up2wing the PR looks good to me. Last thing is about the integration test that could be slightly extended to test a bit more than just if `ch-remote` returns an error or not.
> 
> @rbradford PTAL

The elf check in integration test have been added, please take a review. Thx. @sboeuf

sboeuf reviewed (approved):
Looks ready to me.

up2wing commented:
> @up2wing the quality checks are complaining because you forgot to run `cargo fmt --all`.

Oh yes, forgive my can't wait mood :-)
Fixed.

sboeuf commented:
@rbradford could you please review this PR?

rbradford commented on vmm/src/api/openapi/cloud-hypervisor.yaml:
```diff
@@ -1110,6 +1128,12 @@ components:
+    VmCoredumpData:
+      type: object
+      properties:
+        filepath:
```
I would name this `destination_uri` so it's clear it's a URL.

up2wing replied:
Okay, thanks. So would you do this when merge or should I rebase another commit? @rbradford

sboeuf replied:
Please fix it directly on this PR, by updating the commit where you introduced the code. (So yes please rebase)

up2wing replied:
Ok, I'll rebase immediately.

up2wing replied:
The parameter and member of structure have been renamed as suggestions. @rbradford @sboeuf

rbradford commented on vmm/src/lib.rs:
```diff
@@ -560,6 +564,15 @@ impl Vmm {
     }
 
+    #[cfg(feature = "guest_debug")]
+    fn vm_coredump(&mut self, filepath: &str) -> result::Result<(), VmError> {
```
nit: ditto here

rbradford commented on vmm/src/api/mod.rs:
```diff
@@ -189,6 +192,12 @@ pub struct VmSnapshotConfig {
+#[derive(Clone, Deserialize, Serialize, Default, Debug)]
+pub struct VmCoredumpData {
+    /// The coredump destination file
+    pub filepath: String,
```
ditto and here.

sboeuf commented:
@up2wing thank you for working hard on this PR and addressing all our comments. I think we can merge it now :)
